kvserver: provide control over the rate zone config updates are proce…

ajwerner opened the pull request:
…ssed

For a schema change heavy workload on a small number of vCPUs this can be
a big deal. It's an especially big deal to give a knob. For the workload
I was looking at, the knob doesn't make a big deal because, at the
default settings, the rate we allow was just about the duration it
took to process an update (as in it took more than 1s). However
if I turn down the setting, then I can pretty much completely
eliminate the overhead.

As far as the release note goes, I don't know, I'm putting something but
these low-level knobs are controversial.

Release note (ops change): Added a knob to control the rate at which schema
changes are processed by the key-value store to determine whether their
configuration has changed.

ajwerner committed:
kvserver: provide control over the rate zone config updates are processed

For a schema change heavy workload on a small number of vCPUs this can be
a big deal. It's an especially big deal to give a knob. For the workload
I was looking at, the knob doesn't make a big deal because, at the
default settings, the rate we allow was just about the duration it
took to process an update (as in it took more than 1s). However
if I turn down the setting, then I can pretty much completely
eliminate the overhead.

As far as the release note goes, I don't know, I'm putting something but
these low-level knobs are controversial.

Release note (ops change): Added a knob to control the rate at which schema
changes are processed by the key-value store to determine whether their
configuration has changed.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/62286)
<!-- Reviewable:end -->

ajwerner committed:
kvserver: do not spawn a goroutine per range

Before this change, the code would spawn a goroutine per range when
attempting to add to the merge and split queue. There was a limit on
the parallelism, but it was high (20). Furthermore, we'd end up growing
the stack for each and every goroutine which ends up being expensive.

Release note: None
